community/kernel/src/main/java/org/neo4j/kernel/impl/transaction/state/RecoveryVisitor.java

```diff
@@ -54,11 +62,14 @@ public RecoveryVisitor( TransactionIdStore store,
     public boolean visit( CommittedTransactionRepresentation transaction ) throws IOException
     {
         long txId = transaction.getCommitEntry().getTxId();
-        try ( LockGroup locks = new LockGroup() )
+        TransactionRepresentation txRepresentation = transaction.getTransactionRepresentation();
+
+        try ( LockGroup locks = new LockGroup();
+              ValidatedIndexUpdates indexUpdates = indexUpdatesValidator.validate( txRepresentation, RECOVERY ) )
```

Comment:
Correct, recovery does work under condition that all updates are valid and this stage here is needed to infer `NodePropertyUpdate` objects from commands and group them by index. Will add comment.